fix: sets previousValue for afterChange hook in nested fields correctly

ysfaran opened the pull request:
Fixes: #4643
## Description

<!-- Please include a summary of the pull request and any related issues it fixes. Please also include relevant motivation and context. -->

- [x] I have read and understand the [CONTRIBUTING.md](https://github.com/payloadcms/payload/blob/main/CONTRIBUTING.md) document in this repository.

## Type of change

<!-- Please delete options that are not relevant. -->

- [ ] Chore (non-breaking change which does not add functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Change to the [templates](https://github.com/payloadcms/payload/tree/main/templates) directory (does not affect core functionality)
- [ ] Change to the [examples](https://github.com/payloadcms/payload/tree/main/examples) directory (does not affect core functionality)
- [ ] This change requires a documentation update

## Checklist:

- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] Existing test suite passes locally with my changes
  --> only `test\plugin-search\int.spec.ts` fails locally but this seems to be a setup issue, as this doesn't seem to be related to my changes. The error is: `Cannot find module '../../packages/payload/dist/utilities/wait' from 'test/plugin-search/int.spec.ts'` 
- [ ] I have made corresponding changes to the documentation

DanRibbens commented on packages/payload/src/fields/hooks/afterChange/promise.ts:
```diff
@@ -105,7 +105,8 @@ export const promise = async ({
               previousDoc,
-              previousSiblingDoc: previousDoc?.[field.name]?.[i] || ({} as Record<string, unknown>),
+              previousSiblingDoc:
+                previousSiblingDoc?.[field.name]?.[i] || ({} as Record<string, unknown>),
```
Is this expected?
Imagine a user reorders items in an array and then resaves it?
The position in the array has changed so the previous data just won't line up correctly for the hook to work as expected.

Shouldn't we `find` by the `id` (internal) to the array so that we can have the hook's previous value be more easily used?

ysfaran replied:
I actually had that case in my project and it's kinda confusing, yes. In my case I then used the `afterChange` hook on the array field itself to do the comparison manually.

Does an array item always have an (internal) `id` in payload?

For example, would this ever be possible:

```ts
{
  array: [
    "item1",
    "item2",
    "item3"
  ]
}
```

or is an array item always an object with at least an `id` property:

```ts
{
  array: [
    { id: "id1", text: "item1" }
    { id: "id2", text: "item1" }
    { id: "id3", text: "item1" }
  ]
}
```

If yes, I absolutely agree and will change the code accordingly.

DanRibbens replied:
Yes, by the time the operations get to the `afterChange` hook, you can expect the items to always have an `id`. I believe this happens in the `beforeChange` hook.

Also, we have a function already called `getExistingRowDoc` which might be what you need to use.

ysfaran replied:
@DanRibbens i just pushed another commit to address this issue.

DanRibbens reviewed (changes requested):
This looks good overall, I just have one question to consider.

DanRibbens reviewed (changes requested):
I really appreciate your work here. Sadly, I realized that this same logic for `array` type also needs to be applied to the `blocks` type too.

Are you able to make do this change real quick?

From a testing standpoint as long as this is the same logic as arrays I don't think we necessarily need the extra test case for blocks if we want to skip that.

I realize I'm being annoying in forgetting this in my last review :/

ysfaran commented:
No worries, I'm happy to contribute to this amazing project and appreciate all the work and efforts you guys put into payload. (:

This being said, I hope to create another commit this weekend.

ysfaran commented:
@DanRibbens I added a small commit now.

I also deleted the block field test entirely and hope this is what you meant.

Please let know if there is something else missing.

ysfaran commented:
@DanRibbens what's the status on this? Is something blocking this from getting merged?

ysfaran commented:
@DanRibbens just a reminder.

chrisatcdm commented:
This is a pesky one - I spent way more time than I'll admit experiencing this issue last evening. It applies to nested afterChange hooks, and even in tabs. The previousValue just doesn't behave as expected.

Angusross111 commented:
Any update on this? Im still seeing it in v3.50.0
